File: win-prog/ProgWinVB/Buttons and Labels and Scrolls (Oh, My!)/ColorTrackBar/ColorTrackBar.vb
'----------------------------------------------
' ColorTrackBar.vb (c) 2002 by Charles Petzold
'----------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class ColorTrackBar
    Inherits Form

    Private pnl As Panel
    Private alblName(2) As Label
    Private alblValue(2) As Label
    Private atrkbar(2) As TrackBar

    Shared Sub Main()
        Application.Run(New ColorTrackBar())
    End Sub

    Sub New()
        Text = "Color Track Bar"
        Dim aclr As Color() = {Color.Red, Color.Green, Color.Blue}

        ' Create the panel.
        pnl = New Panel()
        pnl.Parent = Me
        pnl.Location = New Point(0, 0)
        pnl.BackColor = Color.White

        ' Loop through the three colors.
        Dim i As Integer
        For i = 0 To 2
            alblName(i) = New Label()
            alblName(i).Parent = pnl
            alblName(i).ForeColor = aclr(i)
            alblName(i).Text = "&" & aclr(i).ToKnownColor().ToString()
            alblName(i).TextAlign = ContentAlignment.MiddleCenter

            atrkbar(i) = New TrackBar()
            atrkbar(i).Parent = pnl
            atrkbar(i).Orientation = Orientation.Vertical
            atrkbar(i).BackColor = aclr(i)
            atrkbar(i).SmallChange = 1
            atrkbar(i).LargeChange = 16
            atrkbar(i).Minimum = 0
            atrkbar(i).Maximum = 255
            atrkbar(i).TickFrequency = 16
            AddHandler atrkbar(i).ValueChanged, _
                                        AddressOf TrackBarOnValueChanged

            alblValue(i) = New Label()
            alblValue(i).Parent = pnl
            alblValue(i).TextAlign = ContentAlignment.MiddleCenter
        Next i

        Dim clr As Color = BackColor
        atrkbar(0).Value = clr.R           ' Generates ValueChanged event.
        atrkbar(1).Value = clr.G
        atrkbar(2).Value = clr.B

        OnResize(EventArgs.Empty)
    End Sub

    Protected Overrides Sub OnResize(ByVal ea As EventArgs)
        MyBase.OnResize(ea)

        Dim cx As Integer = ClientSize.Width
        Dim cy As Integer = ClientSize.Height
        Dim cyFont As Integer = Font.Height
        Dim i As Integer

        pnl.Size = New Size(cx \ 2, cy)

        For i = 0 To 2
            alblName(i).Location = New Point(i * cx \ 6, cyFont \ 2)
            alblName(i).Size = New Size(cx \ 6, cyFont)
            atrkbar(i).Height = cy - 4 * cyFont
            atrkbar(i).Location = _
                New Point((1 + 2 * i) * cx \ 12 - atrkbar(i).Width \ 2, _
                          2 * cyFont)
            alblValue(i).Location = New Point(i * cx \ 6, _
                                              cy - 3 * cyFont \ 2)
            alblValue(i).Size = New Size(cx \ 6, cyFont)
        Next i
    End Sub

    Private Sub TrackBarOnValueChanged(ByVal obj As Object, ByVal ea As EventArgs)
        Dim i As Integer
        For i = 0 To 2
            If obj Is atrkbar(i) Then
                alblValue(i).Text = atrkbar(i).Value.ToString()
            End If
        Next i
        BackColor = Color.FromArgb(atrkbar(0).Value, _
                                   atrkbar(1).Value, _
                                   atrkbar(2).Value)
    End Sub
End Class

File: win-prog/ProgWinCS/Buttons and Labels and Scrolls (Oh, My!)/ColorTrackBar/ColorTrackBar.cs
//--------------------------------------------
// ColorTrackBar.cs © 2001 by Charles Petzold
//--------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class ColorTrackBar: Form
{
     Panel      panel;
     Label[]    alabelName  = new Label[3];
     Label[]    alabelValue = new Label[3];
     TrackBar[] atrackbar   = new TrackBar[3];

     public static void Main()
     {
          Application.Run(new ColorTrackBar());
     }
     public ColorTrackBar()
     {
          Text = "Color Track Bar";

          Color[] acolor = { Color.Red, Color.Green, Color.Blue };

               // Create the panel.

          panel = new Panel();
          panel.Parent = this;
          panel.Location = new Point(0, 0);
          panel.BackColor = Color.White;

               // Loop through the three colors.

          for (int i = 0; i < 3; i++)
          {
               alabelName[i] = new Label();
               alabelName[i].Parent = panel;
               alabelName[i].ForeColor = acolor[i];
               alabelName[i].Text = "&" + acolor[i].ToKnownColor();
               alabelName[i].TextAlign = ContentAlignment.MiddleCenter;

               atrackbar[i] = new TrackBar();
               atrackbar[i].Parent = panel;
               atrackbar[i].Orientation = Orientation.Vertical;
               atrackbar[i].BackColor = acolor[i];
               atrackbar[i].SmallChange = 1;
               atrackbar[i].LargeChange = 16;
               atrackbar[i].Minimum = 0;
               atrackbar[i].Maximum = 255;
               atrackbar[i].TickFrequency = 16;
               atrackbar[i].ValueChanged += 
                              new EventHandler(TrackBarOnValueChanged);

               alabelValue[i] = new Label();
               alabelValue[i].Parent = panel;
               alabelValue[i].TextAlign = ContentAlignment.MiddleCenter;
          }
          Color color = BackColor;
          atrackbar[0].Value = color.R; // Generates ValueChanged event.
          atrackbar[1].Value = color.G;
          atrackbar[2].Value = color.B;

		  OnResize(EventArgs.Empty);
     }
     protected override void OnResize(EventArgs ea)
     {
          base.OnResize(ea);

          int cx = ClientSize.Width;
          int cy = ClientSize.Height;
          int cyFont = Font.Height;

          panel.Size = new Size(cx / 2, cy);

          for (int i = 0; i < 3; i++)
          {
               alabelName[i].Location = new Point(i * cx / 6, cyFont / 2);
               alabelName[i].Size = new Size(cx / 6, cyFont);

               atrackbar[i].Height = cy - 4 * cyFont;
               atrackbar[i].Location = 
                    new Point((1 + 2 * i) * cx / 12 - atrackbar[i].Width / 2,
                              2 * cyFont);

               alabelValue[i].Location = new Point(i * cx / 6,
                                                   cy - 3 * cyFont / 2);
               alabelValue[i].Size = new Size(cx / 6, cyFont);
          }
     }
     void TrackBarOnValueChanged(object obj, EventArgs ea)
     {
          for (int i = 0; i < 3; i++)
               if((TrackBar) obj == atrackbar[i])
                    alabelValue[i].Text = atrackbar[i].Value.ToString();

          BackColor = Color.FromArgb(atrackbar[0].Value, 
                                     atrackbar[1].Value,
                                     atrackbar[2].Value);
     }
}